src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -464,68 +462,69 @@ public NetworkPeer(NetworkAddress peerAddress, Network network, NetworkPeerConne
 
-                            // Otherwise wait for the socket connection to complete OR if the operation got cancelled.
-                            WaitHandle.WaitAny(new WaitHandle[] { completedEvent, parameters.ConnectCancellation.WaitHandle });
+                        // If the socket connected straight away (synchronously) unblock all threads.
+                        if (!socket.ConnectAsync(nodeSocketEventManager.SocketEvent))
+                            completedEvent.Set();
 
-                            parameters.ConnectCancellation.ThrowIfCancellationRequested();
+                        // Otherwise wait for the socket connection to complete OR if the operation got cancelled.
+                        WaitHandle.WaitAny(new WaitHandle[] { completedEvent, parameters.ConnectCancellation.WaitHandle });
 
-                            if (nodeSocketEventManager.SocketEvent.SocketError != SocketError.Success)
-                                throw new SocketException((int)nodeSocketEventManager.SocketEvent.SocketError);
+                        parameters.ConnectCancellation.ThrowIfCancellationRequested();
 
-                            var remoteEndpoint = (IPEndPoint)(socket.RemoteEndPoint ?? nodeSocketEventManager.SocketEvent.RemoteEndPoint);
-                            this.RemoteSocketAddress = remoteEndpoint.Address;
-                            this.RemoteSocketEndpoint = remoteEndpoint;
-                            this.RemoteSocketPort = remoteEndpoint.Port;
+                        if (nodeSocketEventManager.SocketEvent.SocketError != SocketError.Success)
+                            throw new SocketException((int)nodeSocketEventManager.SocketEvent.SocketError);
 
-                            this.State = NetworkPeerState.Connected;
-                            this.ConnectedAt = DateTimeOffset.UtcNow;
+                        var remoteEndpoint = (IPEndPoint)(socket.RemoteEndPoint ?? nodeSocketEventManager.SocketEvent.RemoteEndPoint);
+                        this.RemoteSocketAddress = remoteEndpoint.Address;
+                        this.RemoteSocketEndpoint = remoteEndpoint;
+                        this.RemoteSocketPort = remoteEndpoint.Port;
 
-                            NodeServerTrace.Information("Outbound connection successful.");
-                        }
+                        this.State = NetworkPeerState.Connected;
+                        this.ConnectedAt = DateTimeOffset.UtcNow;
+
+                        this.logger.LogTrace("Outbound connection to '{0}' established.", peerAddress.Endpoint);
                     }
                 }
-                catch (OperationCanceledException)
-                {
-                    Utils.SafeCloseSocket(socket);
-                    NodeServerTrace.Information("Connection to node cancelled");
-                    this.State = NetworkPeerState.Offline;
+            }
+            catch (OperationCanceledException)
```

Comment:
How about logging what;s in the error object?

Reply:
sorry i don't understand what you mean
when exception happens and it is not cancellation, it will be that catch Exception case bellow which logs the full exception
i fail to see what is that you call "error object"

Reply:
I meant the exception like in ` catch (OperationCanceledException ex)` and do `ex.Message` or something.

Reply:
i've added the log for that but not the message because i would expect operation cancelled exception to just tell me that the operation was cancelled, i'm not sure there any valuable information inside that message

Reply:
How can you tell that there is no useful info in that message?
Is it not better to log it, just in case?